### ISSUE TITLE
cppcoreguidelines-rvalue-reference-param-not-moved false positive on dependent type

### ISSUE THREAD

**Issue by jordan-woyak** [llvm/llvm-project#164735]
The following code demonstrates the issue:
```cpp
#include <memory>
#include <utility>

template <typename T>
struct Holder
{
    std::decay_t<T> value;
    explicit Holder(T&& v) : value{std::forward<T>(v)} {};
};

template <typename T>
Holder(T&&) -> Holder<T>;

int main()
{
    std::unique_ptr<int> x{};
    Holder uptr_holder(std::move(x));

    int y{};
    Holder int_holder(y);
}

```

This produces a clang-tidy warning on the `Holder` constructor:
```
Rvalue reference parameter 'v' is never moved from inside the function body clang-tidy(cppcoreguidelines-rvalue-reference-param-not-moved)
```

`cppcoreguidelines-rvalue-reference-param-not-moved` fails to recognize that `T` may be something like `int&`, making `T&&` also `int&`, and not actually an rvalue reference.

FYI: Replacing `explicit Holder(T&& v)` with `explicit Holder(std::add_rvalue_reference_t<T> v)` silences the warning.
